### PR TITLE
chore: add tasks data layer

### PR DESCRIPTION
**Pull request by ReeceHoffmann**
Changes

 - Converts functions from tasks/pg.py to be part of new tasks datalayer.
   - Updates previous uses of tasks/pg.py to fetch and use datalayer from `app`
 - Updates tests as needed to check for correct datatypes

### CODE REVIEW

File: virtool/tasks/data.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+import virtool.utils
+from virtool.data.errors import ResourceNotFoundError
+from virtool.tasks.models import Task
+
+
+class TasksData:
+    def __init__(self,  pg: AsyncEngine):
+        self._pg = pg
+
+    async def find(self) -> list:
```
Comment (igboyes) [inline virtool/tasks/data.py:15]:
```suggestion
    async def find(self) -> List[Task]:
```

File: virtool/tasks/data.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+            created_at=virtool.utils.timestamp(),
+            progress=0,
+            type=task_class.task_type,
+        )
+        async with AsyncSession(self._pg) as session:
+            session.add(task)
+            await session.flush()
+            document = task.to_dict()
+            await session.commit()
+        return TaskModel(**document)
```
Comment (igboyes) [inline virtool/tasks/data.py:69]: Use blank lines to break up ideas.

```suggestion
        task = Task(
            complete=False,
            context=context or {},
            step=task_class.task_type,
            count=0,
            created_at=virtool.utils.timestamp(),
            progress=0,
            type=task_class.task_type,
        )
        
        async with AsyncSession(self._pg) as session:
            session.add(task)
            await session.flush()
            document = task.to_dict()
            await session.commit()
            
        return TaskModel(**document)
```

**Review by igboyes — CHANGES_REQUESTED** [review]

**Comment by ReeceHoffmann** [comment]
Force pushed to rebase onto main

File: virtool/analyses/data.py [hunk 3]
```diff
@@ -347,7 +347,7 @@ async def blast(self, analysis_id: str, sequence_index: int) -> Optional[str]:
             session.add(blast)
             await session.flush()
 
-            await self._tasks.add(
+            await self._tasks_data.add(
```
Comment (igboyes) [inline virtool/analyses/data.py:350]:
```suggestion
            await self.data.tasks.add(
```
Remove `self._tasks_data` as well.

### Explanation
https://github.com/virtool/virtool/blob/fdb5695bfe273ef346b7d55cfe116a7372486152/virtool/data/piece.py#L10
https://github.com/virtool/virtool/blob/fdb5695bfe273ef346b7d55cfe116a7372486152/virtool/data/layer.py#L41

File: virtool/blast/data.py [hunk 4]
```diff
@@ -16,10 +16,10 @@ class BLASTData:
     A data layer domain for NuVs BLAST data.
     """
 
-    def __init__(self, db, pg: AsyncEngine, tasks: TasksClient):
+    def __init__(self, db, pg: AsyncEngine, tasks_data: TasksData):
         self._db = db
         self._pg = pg
-        self._tasks = tasks
+        self._tasks_data = tasks_data
```
Comment (igboyes) [inline virtool/blast/data.py:22]: Use `self.data` like in analyses. You will have to call `bind_layer`. That is not in `__post_init__` yet.

Do this for all further situations.

File: virtool/references/api.py [hunk 5]
```diff
@@ -188,7 +188,7 @@ async def update(req):
         "user_id": user_id,
     }
 
-    task = await req.app["tasks"].add(UpdateRemoteReferenceTask, context=context)
+    task = await get_data_from_req(req).tasks.add(UpdateRemoteReferenceTask, context=context)
```
Comment (igboyes) [inline virtool/references/api.py:191]: Let's take the opportunity to change this from `add` to `create` to match the rest of the data layer.

File: virtool/startup.py [hunk 6]
```diff
@@ -423,17 +419,17 @@ async def startup_tasks(app: Application):
...
-    await app["tasks"].add(MoveSampleFilesTask)
-    await app["tasks"].add(CleanReferencesTask)
+    await get_data_from_app(app).tasks.add(AddIndexFilesTask)
+    await get_data_from_app(app).tasks.add(AddIndexJSONTask)
+    await get_data_from_app(app).tasks.add(DeleteReferenceTask, context={"user_id": "virtool"})
+    await get_data_from_app(app).tasks.add(AddSubtractionFilesTask)
+    await get_data_from_app(app).tasks.add(StoreNuvsFilesTask)
+    await get_data_from_app(app).tasks.add(CompressSamplesTask)
+    await get_data_from_app(app).tasks.add(MoveSampleFilesTask)
+    await get_data_from_app(app).tasks.add(CleanReferencesTask)
```
Comment (igboyes) [inline virtool/startup.py:433]: Can just call `get_data_from_app` once here and store the reference.

File: virtool/tasks/data.py [hunk 7]
```diff
@@ -141,3 +145,43 @@ async def remove(self, task_id: int):
 
             session.delete(task)
             await session.commit()
+
+    async def add(self, task_class: Type[Task], context: dict = None):
```
Comment (igboyes) [inline virtool/tasks/data.py:149]: ### Naming
Rename `add()` to `create()` to realign with rest of data layer.

### Use `Optional`
```suggestion
    async def add(self, task_class: Type[Task], context: Optional[Dict] = None):
```

https://docs.python.org/3/library/typing.html#typing.Optional

**Review by igboyes — CHANGES_REQUESTED** [review]

**Review by igboyes — APPROVED** [review]

**Pull request merged**